packages/x-date-pickers/src/internals/hooks/usePickerState.ts

```diff
@@ -41,127 +79,162 @@ interface PickerStateProps<TInput, TDateValue> {
+    committed: parsedDateValue,
+    draft: parsedDateValue,
+    resetTarget: parsedDateValue,
+  }));
 
-  // Mobile keyboard view is a special case.
-  // When it's open picker should work like closed, because we are just showing text field
-  const [isMobileKeyboardViewOpen, setMobileKeyboardViewOpen] = React.useState(false);
+  const setDate = React.useCallback(
+    (params: DateStateAction<TDate>) => {
+      setDateState((prev) => {
+        switch (params.action) {
+          case 'setAll':
+          case 'acceptAndClose': {
+            return { draft: params.value, committed: params.value, resetTarget: params.value };
+          }
+          case 'setCommitted': {
+            return { ...prev, draft: params.value, committed: params.value };
+          }
+          case 'setDraft': {
+            return { ...prev, draft: params.value };
+          }
+          default: {
+            return prev;
+          }
+        }
+      });
 
-  const acceptDate = React.useCallback(
-    (acceptedDate: TDateValue, needClosePicker: boolean) => {
-      onChange(acceptedDate);
+      if (!params.internal) {
+        onChange(params.value);
+      }
 
-      if (needClosePicker) {
+      if (params.action === 'acceptAndClose') {
         setIsOpen(false);
-        setInitialDate(acceptedDate);
         if (onAccept) {
-          onAccept(acceptedDate);
+          onAccept(params.value);
         }
       }
     },
     [onAccept, onChange, setIsOpen],
   );
 
+  React.useEffect(() => {
+    if (parsedDateValue != null) {
+      setLastValidDateValue(parsedDateValue);
+    }
+  }, [parsedDateValue]);
+
+  // Set the draft and committed date to equal the new prop value.
+  if (!valueManager.areValuesEqual(utils, dateState.committed, parsedDateValue)) {
+    setDate({ action: 'setCommitted', value: parsedDateValue, internal: true });
+  }
+
+  const shouldCloseOnSelect = !(disableCloseOnSelect ?? wrapperVariant === 'mobile');
+
   const wrapperProps = React.useMemo(
     () => ({
       open: isOpen,
-      onClear: () => acceptDate(valueManager.emptyValue, true),
-      onAccept: () => acceptDate(draftState.draft, true),
-      onDismiss: () => acceptDate(initialDate, true),
+      onClear: () => {
+        // Reset all date in state to the empty value and close picker
+        setDate({ value: valueManager.emptyValue, action: 'acceptAndClose' });
+      },
+      onAccept: () => {
+        // Set all date in state to equal the current draft value and close picker
+        setDate({ value: dateState.draft, action: 'acceptAndClose' });
+      },
+      onDismiss: () => {
+        if (shouldCloseOnSelect) {
+          // Set all dates in state to equal the last accepted date
+          // e.g. Reset the state to the last accepted value
+          setDate({ value: dateState.resetTarget, action: 'acceptAndClose' });
+        } else {
+          // Set all dates in state to equal the last committed date
```

Comment:
```suggestion
          // Set input value to the committed one
          // Equivalent to resetting if no value entered and validating if valid value entere
```

I'm wondering, is it possible for the user to enter a date that will be committed and then modify the input such that the new displayed value is not valid. If it's possible we should check that `draft` is the same date as `committed`

Reply:
I am not following you here
From what I understand, when you edit the input, we only fire the `onChange` when the date is valid (by definition since we call `onChange` with a date)
So `props.value` always contain a valid date.

Do you mean "not valid" as "not a valid date" or as "a date that make no sense" (for instance having year '19' while typing 1997) ?

Reply:
Ok, I was inventing problems that can not occur 😅